substrate-archive/src/database/listener.rs

```diff
@@ -158,22 +166,24 @@ where
 					complete => break,
 				};
 			}
+
 			// collect the rest of the results, before exiting, as long as the collection completes
 			// in a reasonable amount of time
-			let timeout = Delay::new(Duration::from_secs(1));
-			futures::select! {
-				_ = FutureExt::fuse(timeout) => {},
-				notifs = listener.collect::<Vec<_>>().fuse() => {
-					for msg in notifs {
-						self.handle_listen_event(msg.unwrap(), &mut conn).await;
-					}
+			let gather_unfinished = || async {
+				let notifs = listener.collect::<Vec<_>>().await;
+				for msg in notifs {
+					self.handle_listen_event(msg?, &mut conn).await;
 				}
```

Comment:
Could you use `listener.for_each()` here you think?

Reply:
Tried this but because we need a `&mut reference` for the Postgres connection so we can't use that inside of a closure here + it's easier to handle the errors this way (for_each would force us to return a `()`)